Funnel transformer

sgugger opened the pull request:
<!-- This line specifies which issue to close after the pull request is merged. -->
Fixes #4844 

This PR adds the Funnel Transformer architecture in PyTorch. For now, I have uploaded two of the ten checkpoints for this model, I will convert and upload the other ones while this PR is under review and add them before it's merged.

Due to the fact there are two versions of the Funnel model (one that returns hidden states with a sequence length divided by 4 and one that return hidden states with the same sequence length, that has 2 more layers), I add to make to different Tester in the test file (because the expected number of hidden states / attentions change depending on which model is used). I adapted the script that check all models are tested to account for that.

sgugger commented on src/transformers/modeling_funnel.py:
```diff
@@ -0,0 +1,1528 @@
+            if output_attentions:
+                idx += 1
+                outputs = outputs + (encoder_outputs[2] + decoder_outputs[idx],)
+            return outputs
```
I know this bit is ugly but I didn't have anything better. Any suggestions to make it less horrible are welcome!

patrickvonplaten replied:
Shouldn't it be `outputs = outputs + (encoder_outputs[idx] + decoder_outputs[idx],)` instead of `outputs = outputs + (encoder_outputs[1] + decoder_outputs[idx],)`? 

I think the code as it is now is not correct if `output_hidden_states` is False and `output_attentions` is True or am I missing something?

patrickvonplaten replied:
In T5 and Bart we actually just do `encoder_outputs` + `decoder_outputs` instead of grouping the outputs, but it's questionable whether this is the better way to do it...

sgugger replied:
No, the encoder has a hard-coded `output_hidden_states=True` since we need the last hidden state of the first block for the upsampling.

sgugger commented on src/transformers/tokenization_funnel.py:
```diff
@@ -0,0 +1,243 @@
+            return len(cls) * [self.cls_token_type_id] + len(token_ids_0 + sep) * [0]
+        return len(cls) * [self.cls_token_type_id] + len(token_ids_0 + sep) * [0] + len(token_ids_1 + sep) * [1]
+
+    def _convert_encoding(self, encoding, **kwargs):
```
This part is a bit hacky but necessary to make the fast tokenizer work. Basically, the Funnel Transformer authors put 2 in type_token_ids for the cls token (instead of 0) but we can't ask the rust tokenizer to do that for us, so we replace that token_id in the post-processing function.

sgugger commented on tests/test_modeling_common.py:
```diff
@@ -539,7 +539,10 @@ def check_hidden_states_output(inputs_dict, config, model_class):
 
-            self.assertEqual(len(hidden_states), self.model_tester.num_hidden_layers + 1)
+            expected_num_layers = getattr(
+                self.model_tester, "expected_num_hidden_layers", self.model_tester.num_hidden_layers + 1
```
Added this part because the full funnel transformer actually adds two hidden states, not just one.

patrickvonplaten replied:
One for encoder / one for decoder or is this something else?

patrickvonplaten replied:
Ah yeah got it - nvm

LysandreJik commented on docs/source/model_doc/funnel.rst:
```diff
@@ -0,0 +1,126 @@
+Overview
+~~~~~~~~~~~~~~~~~~~~~
+
+The Funnel Transformer model was proposed in the paper.
```
```suggestion
The Funnel Transformer model was proposed in the paper
```

LysandreJik commented on docs/source/model_doc/funnel.rst:
```diff
@@ -0,0 +1,126 @@
+The Funnel Transformer model was proposed in the paper.
+`Funnel-Transformer: Filtering out Sequential Redundancy for Efficient Language Processing
+<https://arxiv.org/abs/2006.03236>`__.
+It is a bidirectional transformer model, like BERT, but with a pooling operation after each blocks of layers, a bit
```
```suggestion
It is a bidirectional transformer model, like BERT, but with a pooling operation after each block of layers, a bit
```

LysandreJik commented on docs/source/model_doc/funnel.rst:
```suggestion
- Since Funnel Transformer uses pooling, the sequence length of the hidden states changes after each block of layers.
```

LysandreJik commented on src/transformers/configuration_funnel.py:
```diff
@@ -0,0 +1,184 @@
+FUNNEL_PRETRAINED_CONFIG_ARCHIVE_MAP = {
+    "funnel-small": "https://s3.amazonaws.com/models.huggingface.co/bert/huggingface/funnel-small/config.json",
+    "funnel-small-base": "https://s3.amazonaws.com/models.huggingface.co/bert/huggingface/funnel-small-base/config.json",
+    # "funnel-medium": "",
+    # "funnel-medium-base": "",
+    # "funnel": "",
+    # "funnel-base": "",
+    # "funnel-large": "",
+    # "funnel-large-base": "",
+    # "funnel-xlarge": "",
+    # "funnel-xlarge-base": "",
```
Should this be uncommented/filled?

sgugger replied:
Yes, when I have all models on the S3.

LysandreJik commented on src/transformers/configuration_funnel.py:
```diff
@@ -0,0 +1,184 @@
+
+    Args:
+        vocab_size (:obj:`int`, `optional`, defaults to 30522):
+            Vocabulary size of the BERT model. Defines the different tokens that
+            can be represented by the `inputs_ids` passed to the forward method of :class:`~transformers.BertModel`.
```
```suggestion
            Vocabulary size of the Funnel transformer. Defines the different tokens that
            can be represented by the `inputs_ids` passed to the forward method of :class:`~transformers.FunnelModel`.
```

LysandreJik commented on src/transformers/convert_funnel_original_tf_checkpoint_to_pytorch.py:
```diff
@@ -0,0 +1,61 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Convert XXX checkpoint."""
```
```suggestion
"""Convert Funnel checkpoint."""
```

LysandreJik commented on src/transformers/modeling_funnel.py:
```diff
@@ -0,0 +1,1529 @@
+FUNNEL_PRETRAINED_MODEL_ARCHIVE_LIST = [
+    "funnel_small",  # B4-4-4H768
+    "funnel_small-base",  # B4-4-4H768, no decoder
+    # "funnel-medium",      # B6-3x2-3x2H768
+    # "funnel-medium-base", # B6-3x2-3x2H768, no decoder
+    # "funnel",             # B6-6-6H768
+    # "funnel-base",        # B6-6-6H768, no decoder
+    # "funnel-large",       # B8-8-8H1024
+    # "funnel-large-base",  # B8-8-8H1024, no decoder
+    # "funnel-xlarge-base", # B10-10-10H1024
+    # "funnel-xlarge",      # B10-10-10H1024, no decoder
+]
```
Same here, we could uncomment those?

LysandreJik commented on src/transformers/modeling_funnel.py:
```diff
@@ -0,0 +1,1529 @@
+
+class FunnelAttentionStructure(nn.Module):
+    """
+    Contains helpers for `FunnelTfmRelMultiheadAttention`.
```
```suggestion
    Contains helpers for `FunnelRelMultiheadAttention `.
```

LysandreJik commented on src/transformers/modeling_funnel.py:
```diff
@@ -0,0 +1,1529 @@
+
+
+@add_start_docstrings(
+    "The bare base Funnel Transformer Model transformer outputting raw hidden-states without any specific head on top.",
```
`The bare base Funnel Transformer Model transformer` I don't understand what that means and I feel like my teeth are going to fall out after reading this out loud

LysandreJik commented on src/transformers/modeling_funnel.py:
```diff
@@ -0,0 +1,1529 @@
+                outputs = outputs + (encoder_outputs[2] + decoder_outputs[idx],)
+            return outputs
+
+        return BaseModelOutput(
+            last_hidden_state=decoder_outputs[0],
+            hidden_states=(encoder_outputs.hidden_states + decoder_outputs.hidden_states)
+            if output_hidden_states
+            else None,
+            attentions=(encoder_outputs.attentions + decoder_outputs.attentions) if output_attentions else None,
+        )
```
Here it seems that you're having the `model_output.hidden_states` and `model_output.attentions` contain both the encoder and decoder outputs, concatenated?

In other encoder/decoder models (T5, BART) or multi-transformer models (LXMERT), we prefer to have separate outputs for those, instead of grouping the values under the same output. Wouldn't it be better to have it be the same here?

It would also simplify the way the output is done if `return_dict=False`

sgugger replied:
It's not a real encoder/decoder model. What they call decoder is just two final layers in the model after upsampling, so the hidden states/attentions should be all together. I can also change the same if that helps (and adapt the conversion script for the weights).

A decoder would follow the same behavior as in the original transformer, which is not the case here.

LysandreJik commented on src/transformers/modeling_funnel.py:
```diff
@@ -0,0 +1,1529 @@
+    Funnel Transformer model with a binary classification head on top as used during pretraining for identifying
+    generated tokens.""",
+    FUNNEL_START_DOCSTRING,
+)
+
+
+class FunnelForPreTraining(FunnelPreTrainedModel):
```
Maybe remove these two lines for readability?
```suggestion
)
class FunnelForPreTraining(FunnelPreTrainedModel):
```

LysandreJik commented on tests/test_modeling_funnel.py:
```diff
@@ -0,0 +1,421 @@
+        use_input_mask=True,
+        use_token_type_ids=True,
+        use_labels=True,
+        vocab_size=30522,
```
That's a big vocab_size for tests isn't it?

sgugger replied:
Oops

LysandreJik reviewed (commented):
This is very cool, very excited to welcome the Funnel Transformer into the library! Looking forward to the TF version!

patrickvonplaten commented on src/transformers/configuration_funnel.py:
```diff
@@ -0,0 +1,184 @@
+# coding=utf-8
+# Copyright 2010, XXX authors
```
```suggestion
# Copyright 2020, Hugging Face 
```

patrickvonplaten commented on src/transformers/convert_funnel_original_tf_checkpoint_to_pytorch.py:
```diff
@@ -0,0 +1,61 @@
+# coding=utf-8
+# Copyright 2018 The HuggingFace Inc. team.
```
```suggestion
# Copyright 2020 The HuggingFace Inc. team.
```

patrickvonplaten commented on src/transformers/modeling_funnel.py:
```diff
@@ -0,0 +1,1529 @@
+
+FunnelLayerNorm = nn.LayerNorm
+
+INF = 1e6
```
usually we use 10000.0 for the attention mask. Do we have to use 10^6 here? I think this could lead to problems with fp16 because 
```
torch.tensor([10**6]).to(torch.float16)
```
yields inf, while 
```
torch.tensor([10**4]).to(torch.float16)
```
does not

sgugger replied:
This will probably change the results with respect to the original implementation though. Maybe not sufficiently for us to care?

sgugger replied:
Side question, why are infs bad since it goes in a softmax anyway?

patrickvonplaten replied:
If it changes the results with the original implementation I think we should keep 10**6 - yeah I guess as long as not all values are `inf` inf that goes into a softmax should be fine (think distillbert) does the same

patrickvonplaten commented on src/transformers/modeling_funnel.py:
```diff
@@ -0,0 +1,1529 @@
+    Contains helpers for `FunnelTfmRelMultiheadAttention`.
+    """
+
+    cls_token_type_id: int = 2
```
Is this better hardcoded here than being a config param?

sgugger replied:
I don't think anyone will ever want to change it so not sure it's worth a config param.

patrickvonplaten commented on src/transformers/modeling_funnel.py:
```diff
@@ -0,0 +1,1529 @@
+        """ Returns the attention structure associated to the inputs. """
+        # input_embdeds has shape batch_size x seq_len x d_model
+        # attention_mask and token_type_ids have shape batch_size x seq_len
+        self.pooling_mult = 1
```
why not directly set it at init?

sgugger replied:
This is called at each new forward, so it needs to be reset in the init.

patrickvonplaten commented on src/transformers/modeling_funnel.py:
```diff
@@ -0,0 +1,1529 @@
+
+        if self.config.separate_cls:
+            t = tensor[:, :-1] if self.config.truncate_seq else tensor
+            tensor = torch.cat([tensor[:, :1], t], dim=1)
```
Out of curiosity, why is the `tensor[:, 0]` value duplicated here?

sgugger replied:
To make sure the state for the cls token is taken independently of the rest.

patrickvonplaten commented on src/transformers/modeling_funnel.py:
```diff
@@ -0,0 +1,1529 @@
+            raise NotImplementedError("The supported modes are 'mean', 'max' and 'min'.")
+
+        if ndim == 2:
+            return tensor[:, 0, :, 0]
```
Is this operation the same as `tensor.squeeze()`?

sgugger replied:
Not if the batch size is 1.

patrickvonplaten replied:
and then we just disregard all other batches?

sgugger replied:
Then the tensor would have a shape [hidden_dim] instead of [1, hidden_dim]

patrickvonplaten commented on src/transformers/modeling_funnel.py:
```diff
@@ -0,0 +1,1529 @@
+            tensor = tensor[:, None, :, None]
+        elif ndim == 3:
+            tensor = tensor[:, None, :, :]
+        stride = (stride, 1)
```
small comment here would be nice to understand why stride dim is added

patrickvonplaten commented on src/transformers/modeling_funnel.py:
```diff
@@ -0,0 +1,1529 @@
+        return attention_struct
+
+
+def _relative_shift_gather(x, context_len, shift):
```
not a big fan of `x` and `y` here -> make it impossible to later rename these variables...maybe just `hidden_states` as in other models.

patrickvonplaten commented on src/transformers/modeling_funnel.py:
```diff
@@ -0,0 +1,1529 @@
+    # # matrix of context_len + i-j
+    # y = x.gather(3, idxs.expand([bs, n_head, context_len, context_len]))
+
+    y = torch.reshape(x, [batch_size, n_head, max_rel_len, seq_len])
```
why introduce new variable `y` here and not stick with `x`?

patrickvonplaten commented on src/transformers/modeling_funnel.py:
```diff
@@ -0,0 +1,1529 @@
+
+    y = torch.reshape(x, [batch_size, n_head, max_rel_len, seq_len])
+    y = y[:, :, shift:, :]
+    # y = torch.narrow(y, 2, shift, max_rel_len - shift)
```
should probably be removed

patrickvonplaten commented on src/transformers/modeling_funnel.py:
```diff
@@ -0,0 +1,1529 @@
+            # phi and pi have shape seq_len x d_model, psi and omega have shape context_len x d_model
+            phi, pi, psi, omega = position_embeds
+            # Shape n_head x d_head
+            u = self.r_r_bias * self.scale
```
what is `u`?

sgugger replied:
Notations of the paper, appendix A.2.2, final formula :-)

sshleifer replied:
link it!

patrickvonplaten commented on src/transformers/modeling_funnel.py:
```diff
@@ -0,0 +1,1529 @@
+            # Shape n_head x d_head
+            u = self.r_r_bias * self.scale
+            # Shape d_model x n_head x d_head
+            w_r = self.r_kernel
```
why not just stick with `self.r_kernel`?

sgugger replied:
Same as above.

patrickvonplaten commented on src/transformers/modeling_funnel.py:
```diff
@@ -0,0 +1,1529 @@
+            shift = 2 if q_head.shape[1] != context_len else 1
+            # Notations from the paper, appending A.2.1, final formula
+            # Grab the proper positional encoding, shape max_rel_len x d_model
+            r = position_embeds[self.block_index][shift - 1]
```
not a fan of all these single letter variables

sgugger replied:
The code can't really be understood without following the paper. I can give funny names to ll those variables, but it will end up less easy to read along the math on the paper.

patrickvonplaten commented on src/transformers/modeling_funnel.py:
```diff
@@ -0,0 +1,1529 @@
+            token_type_attn *= cls_mask
+        return token_type_attn
+
+    def forward(self, q, k, v, attention_struct, head_mask=None, output_attentions=False):
```
would prefer to write `query_input` instead of `q` same for k and v

sgugger replied:
It is obvious from the context that those are inputs, so `query_input` is not a good name. Open to `query`, `key` and `value`, but I'm not sure anyone reading the internal of a complex attention layer really needs any explanation on what q/k/v are.

patrickvonplaten commented on src/transformers/modeling_funnel.py:
```diff
@@ -0,0 +1,1529 @@
+        if attention_mask is not None:
+            attn_score = attn_score - INF * attention_mask[:, None, None].float()
+        # attention probability
+        attn_prob = torch.softmax(attn_score, dim=-1)
```
I think we can see some memory allocation when directly setting the dtype in `torch.softmax(dtype=dtype)`

patrickvonplaten commented on src/transformers/modeling_funnel.py:
```diff
@@ -0,0 +1,1529 @@
+        attn_vec = torch.einsum("bnij,bjnd->bind", attn_prob, v_head)
+
+        # Shape shape batch_size x seq_len x d_model
+        attn_out = self.post_proj(attn_vec.contiguous().view(batch_size, seq_len, n_head * d_head))
```
nit `attn_vec.reshape(...)` is the same as contiguous().view(...) I think

patrickvonplaten commented on src/transformers/modeling_funnel.py:
```diff
@@ -0,0 +1,1529 @@
+        self.dropout = nn.Dropout(config.hidden_dropout)
+        self.layer_norm = FunnelLayerNorm(config.d_model, config.layer_norm_eps)
+
+    def forward(self, x):
```
think `hidden_states` is better than `x`

sgugger replied:
I'll compromise on `hidden` ;-)

patrickvonplaten commented on src/transformers/modeling_funnel.py:
```diff
@@ -0,0 +1,1529 @@
+        return BaseModelOutput(last_hidden_state=hidden, hidden_states=all_hidden_states, attentions=all_attentions)
+
+
+def upsample(x, stride, target_len, separate_cls=True, truncate_seq=False):
```
this function deserves a comment -> very hard to understand what's going on here

patrickvonplaten commented on src/transformers/modeling_funnel.py:
```diff
@@ -0,0 +1,1529 @@
+        input_ids (:obj:`torch.LongTensor` of shape :obj:`{0}`):
+            Indices of input sequence tokens in the vocabulary.
+
+            Indices can be obtained using :class:`transformers.XxxTokenizer`.
```
```suggestion
            Indices can be obtained using :class:`transformers.FunnelTokenizer`.
```

patrickvonplaten commented on src/transformers/modeling_funnel.py:
```diff
@@ -0,0 +1,1529 @@
+        )
+
+        decoder_outputs = self.decoder(
+            encoder_outputs[0],
```
would prefer to use keyword arguments here to better understand what `encoder_outputs[0]` are. (same for following line) => `final_hidden=...` and `final_block_hidden=...`

patrickvonplaten commented on src/transformers/modeling_funnel.py:
```diff
@@ -0,0 +1,1529 @@
+        self.num_labels = config.num_labels
+
+        self.funnel = FunnelBaseModel(config)
+        self.classifier = nn.Sequential(
```
not a huge fan of `nn.Sequential()` here. Colud we maybe create a new `nn.Module` for this? 
I think this keeps the code more flexible and readable

sgugger replied:
I don't see how creating a module makes it more readable, it's actually longer (you have to write the forward pass) while adding nothing of value. I can refactor this in a separate function if needed but I don't really get what the end user will gain in terms of flexibility. In both cases, they'll have to edit this file for modifications.

patrickvonplaten replied:
My arguments are the following (don't have a very strong opinion on this though):

1) Creating a separate model is more in line with how it is done in other models (other models always create a seperate class for things like `classification_head`, `lm_head`, `mlm_head`. Because we never use `nn.Sequential` anywhere in the code, a user is "more" surprised by this new feature -> https://en.wikipedia.org/wiki/Principle_of_least_astonishment

2) IMO, it's less flexible in a sense that it's harder to tweak with special functions, like a residual connection between input first linear layer output or special mathematical functions for which a lambda layer would probably be needed

3) It's impossible to debug between `nn.Linear` and `nn.Tanh()` e.g. if one wants to quickly know the output of the layer, he would have to rewrite the whole `nn.Sequential` object

patrickvonplaten replied:
Okey actually CTLR and Transfo-XL use `nn.Sequential` as well, but I don't think that these are our cleanest models

sgugger replied:
All the arguments are sound, but apply to a complex model. This is just two linear layers with a tanh and a dropout in the middle. Anyhow, will refactor as a module since it's going to take less time than debating about it.

sgugger commented on src/transformers/modeling_funnel.py:
```diff
@@ -0,0 +1,1529 @@
+            return type(tensor)(self.pool_tensor(tensor, mode=mode, stride=stride) for x in tensor)
+
+        if self.config.separate_cls:
+            t = tensor[:, :-1] if self.config.truncate_seq else tensor
```
Just `suffix` though, since it is obvious from the context it is a tensor :-p

patrickvonplaten commented on tests/test_modeling_funnel.py:
```diff
@@ -0,0 +1,421 @@
+        hidden_dropout=0.1,
+        attention_dropout=0.1,
+        activation_dropout=0.0,
+        max_position_embeddings=512,
```
this is also pretty big

sgugger replied:
BertTester uses the same. Do we actually pad to the max in the common tests?

patrickvonplaten replied:
It's just the weights of the position encodings weights that are set according to this parameter, so it's not a big problem I guess. The memory footprint of this weight is not very big

patrickvonplaten commented on tests/test_modeling_funnel.py:
```diff
@@ -0,0 +1,421 @@
+
+    def test_for_multiple_choice(self):
+        config_and_inputs = self.model_tester.prepare_config_and_inputs()
+        self.model_tester.create_and_check_for_multiple_choice(*config_and_inputs)
```
Would be very nice to have hard coded integration tests to make sure future changes to change the output of the model. Think each model should have at least one fast integration test to check that the output slice of a small model gives the expected resultls and one slow integration test.

sgugger replied:
You mean setting all weights manually on a small model and check a given input gives the proper target?

patrickvonplaten replied:
I thought a slow test like this one

https://github.com/huggingface/transformers/blob/5a318f075ad914c58ddca28494e88baabeec242c/tests/test_modeling_longformer.py#L535

=> this makes sure that the pretrained model works (this test should obviously give identical results for the original model)

and also

https://github.com/huggingface/transformers/blob/5a318f075ad914c58ddca28494e88baabeec242c/tests/test_modeling_longformer.py#L332

Such a test can be created at integration time of the model: A hardcoded input is run through the small test funnel transformer model and the output is successively hard coded into the test (because we know that the model works at the time of integration).

patrickvonplaten commented:
Awesome! The model seems quite complex so I didn't really understand all the functionality. 
A couple of things from my side:

1) IMO, it's super useful to have hard  coded integration tests in the test file which makes the model a lot easier to maintain (every change can quickly be checked by making sure the model stays mathematically equivalent).

2) I guess a couple of comments and assert statements would be nice to make the code a bit easier to understand

3) Personally, I don't like single letter variables. Search replace commands don't work on such variables and it is very difficult to understand what they mean.

sshleifer commented on src/transformers/modeling_funnel.py:
```diff
@@ -0,0 +1,1531 @@
+        # dividide.
+        self.pooling_mult = None
+
+    def init_attention_structure(self, input_embeds, attention_mask=None, token_type_ids=None):
```
(nit/naming)
what is an attention_structure?
maybe `get_attention_ingredients`/`get_attention_inputs`?

sgugger replied:
I'd still leave the init instead of get since it's the initial ones, so `init_attention_inputs`.

sshleifer reviewed (commented):
Looks cool, left some nits. +1 to integration testing!

sshleifer commented on src/transformers/modeling_funnel.py:
```diff
@@ -0,0 +1,1531 @@
+
+    def get_position_embeds(self, seq_len, dtype, device):
+        """
+        Create and cache inputs related to relative position encoding.
```
would love to understand what this return type is! Seems like the if and else return very different things.

sshleifer commented on src/transformers/modeling_funnel.py:
```diff
@@ -0,0 +1,1531 @@
+        if isinstance(tensor, (tuple, list)):
+            return type(tensor)(self.pool_tensor(tensor, mode=mode, stride=stride) for x in tensor)
+
+        if self.config.separate_cls:
```
is there a comment about what `separate_cls` controls? It seems powerful!

sgugger replied:
It's documented in the config file. It controls the fact the cls token is ignored in the pooling operations.

sshleifer commented on src/transformers/modeling_funnel.py:
```diff
@@ -0,0 +1,1531 @@
+    # max_rel_len = 2 * context_len + shift -1 is the numbers of possible relative positions i-j
+
+    # What's next is the same as doing the following gather, which might be clearer code but less efficient.
+    # idxs = context_len + torch.arange(0, context_len).unsqueeze(0) - torch.arange(0, context_len).unsqueeze(1)
```
great comment!

sgugger commented on src/transformers/__init__.py:
```diff
@@ -325,6 +327,18 @@
         FlaubertWithLMHeadModel,
     )
+    from .modeling_funnel import (
+        FUNNEL_PRETRAINED_MODEL_ARCHIVE_LIST,
```
Just copying every other model imports in the `__init__`. Might need to be a centralized effort for all models, so should be done in a separate PR if needed.

sgugger commented on tests/test_modeling_funnel.py:
```diff
@@ -0,0 +1,421 @@
+        self.model_tester.create_and_check_for_question_answering(*config_and_inputs)
+
+    @slow
+    def test_model_from_pretrained(self):
```
Note that no checkpoints are usable out of the box to give sensible predictions since it's an ELECTRA-style pretraining. But can add a test of a given output.

sgugger commented:
Thanks for all the comments. I think I replied/addressed all of them except the fast small integration tests, which are going to take a bit more work (starting on this now). Let me know if I missed anything since there are a lot of comments!

patrickvonplaten commented on tests/test_modeling_funnel.py:
```diff
@@ -0,0 +1,454 @@
+
+
+@require_torch
+class FunnelModelIntegrationTest(unittest.TestCase):
```
Awesome!

sgugger commented:
All checkpoints uploaded so I updated the incomplete lists. Also added mention of the model in all indexes, the model summary and the big table of pretrained models (sorry about the diff on that file, Funnel Transformer is one character too long and required to add an extra space on every line).

Should be good to merge at the beginning of next week!

LysandreJik reviewed (approved):
Very cool!

Shashi456 commented:
@sgugger although you've named the models "funnel-base", "funnel-medium" so on so forth, the paper talks about all this in the a different format, could a docstring be added saying `funnel-base` is `B4-4-4H768` and same for the rest. If someone wants to replicate the papers' results that would be great.
